### ISSUE TITLE
Validate disk space available

### ISSUE THREAD

**Issue by jzabroski** [axboe/fio#1169]
**Description of the new feature**
A system call to check available disk prior to allocating all `--numjobs` plus `--size`.

Context: I misunderstood an error complaining about total_size not being specified, and did not realize total_size is `--numjobs * --size`.

**Comment by jzabroski** [comment]
Notes: Obviously, with some file systems, there is no way to know how much space is actually available.  For example, if there is a quota, that may not be reported via a system call that reports free space on the disk.

**Comment by sitsofe** [comment]
Hi @jzabroski,

Trying to validate space has a few design issues:
- How does this work with ioengine that are diskless?
- How do you get past the time of check vs time of use race condition? (There was enough space but then something else gobbled it up before you could get to it)
- There will be interactions with files that already exist because they may be bigger than the size required and have to be deducted from any estimate

Given the complexity and amount of effort to implement I'd say this is difficult but we're open to outside contributions. If you submit an implementation that works everywhere and addresses the corner cases we'll review it!

However you mention an issue with the message you were returned and we can explore that quickly... What sort of text would you suggest?

Further you mention a "hung terminal" - are you using Windows 10? Which terminal are you using?

**Comment by jzabroski** [comment]
> Further you mention a "hung terminal" - are you using Windows 10? Which terminal are you using?

This box had PowerShell 5, so I was using that.
After I had Control+C to stop fio, I observed that terminal was still up and running `Get-Process fio` on another PS terminal showed that the PID was still running but there was no PowerShell process via `Get-Process PowerShell` other than the current terminal I had opened.

In any event, it doesn't seem to make sense that fio continues trying to write bits in randwrite mode when it runs out of disk space to write to.

**Comment by sitsofe** [comment]
@jzabroski

- Windows 10 or some other version of Windows?
- So pressing Ctrl-C didn't actually have an impact?

> it doesn't seem to make sense that fio continues trying to write bits in randwrite mode when it runs out of disk space to write to.

This is interesting. fio has to lay out the file before it starts writing so are you saying the layout was successful but the later writing was not? If so, wouldn't fio have had space to continue writing to the file? I didn't think Windows did sparse files unless you explicitly asked for them - do you know if that's the case? I suppose ancillary files also wouldn't have been writable but the impact depends on just where the lock up happened...

**Comment by jzabroski** [comment]
> * Windows 10 or some other version of Windows?

Windows Server 2019 Data Center Edition .  It's pretty standard for AWS EC2.

> * So pressing Ctrl-C didn't actually have an impact?

I think `taskkill /t /f /im 6902` did the trick, but I had to wait about a minute.  It may also be that fio just timed out since thankfully I only ran the benchmark for 180 seconds.

> I didn't think Windows did sparse files unless you explicitly asked for them - do you know if that's the case?

There is a specific feature in Windows called Instant File Initialization that all database administrators set to true for what I think you're describing when you say "fio has to lay out the file before it starts writing"

**Comment by sitsofe** [comment]
Does Ctrl-C stop fio normally in your terminal?

> Instant File Initialization

(Bear in mind I'm not a Windows person) That looks like an SQL Server feature (which is looks extremely useful for SQL Server admins). What I'd need to know is what does SQL Server actually do on the backend to achieve that feature? I'm guessing (based on the "Security considerations" section) it's the mode where if you have the privileges you tell Windows "Please expose the backing of this file without zeroing it first". but I think that's the `Set­File­Valid­Data()` call (see  https://github.com/axboe/fio/issues/875#issuecomment-566579608 ).

When I say "fio has to lay out the file" I'm referencing the stage where fio tries to preparers the files it wants to do I/O to before it actually starts doing I/O. https://github.com/axboe/fio/blob/fio-3.25/os/os-windows.h doesn't show a Windows  `fio_fallocate()` implementation so there's won't be any special Windows calls to help improve layout but if the file doesn't already exist I would guess fio would manually be growing the file via https://github.com/axboe/fio/blob/fio-3.25/filesetup.c#L212-L241 which should break out if `ENOSPC` is returned. My next question is - was fio in that stage or had it started doing read/write I/O with the ETA counting down etc?

**Comment by jzabroski** [comment]
> What I'd need to know is what does SQL Server actually do on the backend to achieve that feature?

It's not a SQL Server feature, we're just the primary users of the feature.  It's part of `secpol.msc` (Programs ending in msc stand for **M**icrosoft **S**nap-in **Control** and are plug-ins to the Microsoft Management Console), which is the Security Policy editor.  It requires the Perform Volume Maintenance Tasks (a.ka. `SE_MANAGE_VOLUME_NAME`) permission.  This gives the user running the task the ability to manage volumes and disk configurations. 

Side note: There are a couple of neat tricks "SQLOS" does that were best explained in an archaic book, SQL Server 6.5 Internals, written in the 1990s when Microsoft finally bought the software license for Sybase SQL.  Newer authors that cover some internals are Kalen Delaney. -- SQL Server uses this permission for laying out data files, but doesn't do this for log files, because log files have their own "file within file" abstraction managed by SQLOS called virtual log files.  This allows multiple processes to write to the same file, because SQLOS abstracts away the write-able address block for the log writer for a given process's transactions writelog operations.

tl;dr: Instant File Initialization is just a marketing name for managing the volume directly using secure privileges.  It's part of secpol because it's got massive security implications.

![image](https://user-images.githubusercontent.com/447485/104849338-23558a00-58b7-11eb-9243-3abbfb6d1702.png)

![image](https://user-images.githubusercontent.com/447485/104849398-72032400-58b7-11eb-8a55-4736480453a5.png)



> My next question is - was fio in that stage or had it started doing read/write I/O with the ETA counting down etc?

I am believe fio was in that stage because I had 16 files created (for `--numjobs=16) and three of the files had non-0Kb file sizes in Windows Explorer.  I imagine file systems don't do blocking calls when you allocate these files, so if you're not verifying they're allocated by querying the final file size, then that might be the missing step?

**Comment by jzabroski** [comment]
Here is the underlying info of what SQL Server does:

* The OS supports the call to SetFileValidData function ( https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-setfilevaliddata ) - I am not sure if Linux has this, so I am not sure what SQL Server on Linux does, but it looks like [FALLOC_FL_ZERO_RANGE flag for fallocate](https://lwn.net/Articles/586797/) was introduced in kernel 3.15 around ~2014.
* The account (directly or through group membership) has the privilege required
* It is not a sparse file  ( a.k.a. Snapshots)

**Comment by sitsofe** [comment]
As you noted Linux has [fallocate(2)](https://man7.org/linux/man-pages/man2/fallocate.2.html) which achieves the same thing as but in a different style (fio already uses it on Linux). It's different but in its own way convenient (no need for elevated security). It's more about optimal layout though as if you're dong reads by default fio tries to ensure there's random data to read back...

> I imagine file systems don't do blocking calls when you allocate these files

When they are being populated with data in the loop https://github.com/axboe/fio/blob/fio-3.25/filesetup.c#L212-L241 the write call will be blocking but I'd expect the call to fail. What's seems odd is knowing whether ENOSPC was returned (maybe Windows doesn't automatically set `errno` and we'd have to make an explicit check?) and if so where fio went on to get stuck. You would probably need a debugger (e.g. gdb) or lots of manually introduced `log_info()` statements to narrow that one down further. I don't have a Windows machine so I can only offer moral support but if you or someone you know who would like to dig into C please feel free to dig in!

**Comment by jzabroski** [comment]
> https://github.com/axboe/fio/blob/fio-3.25/filesetup.c#L212-L241

I think you need to check for EFBIG as well, looking at https://docs.microsoft.com/en-us/cpp/c-runtime-library/errno-constants?view=msvc-160

**Comment by sitsofe** [comment]
Hmm, I'd be surprised if it was returning EFBIG unless your filesystem was something like FAT32 because [NTFS supports sizes of files up to 8 Petabytes](https://en.wikipedia.org/wiki/NTFS#Scalability) but I guess you never know...

**Comment by jzabroski** [comment]
I may misunderstand that errno, then.  I thought the size of the system was referring to the hardware and not the file system.  Thanks for your patience explaining that to me.

**Comment by jzabroski** [comment]
NTFS disk space free analysis is very hairy. https://docs.microsoft.com/en-us/troubleshoot/windows-server/backup-and-storage/disk-space-problems-on-ntfs-volumes

**Comment by sitsofe** [comment]
@jzabroski Some more questions:

- Can you post the full job/options that resulted in the hang?
- Can you post the error message that you're referred to in your initial comment?
- Is the problem reproducible every time? If so can you make it happen on smaller disks etc?

**Comment by sitsofe** [comment]
Now I think about it some more, I wonder if the "hang" was due to Windows' zero backfilling behaviour... If so you would likely see the issue when you make a huge file and randomly write to it such that a write happens near the end of the file forcing Windows to fill the empty area up to it with zeroes.

**Comment by sitsofe** [comment]
Closing due to lack of reply from reporter. If this issue is still happening with the latest fio (see https://github.com/axboe/fio/releases to find out which version that is) please reopen. Thanks!
